### PR TITLE
Use symbol resolver

### PR DESCRIPTION
**Pull request by ftomassetti**
Using the SymbolResolver in more cases:
* to find declarations
* to get resolved types from types
* to calculate the types of expressions

### CODE REVIEW

File: javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java [hunk 1]
```diff
@@ -22,7 +22,21 @@
...
+    /**
+     * For a reference it would find the corresponding
+     * declaration.
+     */
+    <T> T resolveDeclaration(Node node, Class<T> resultClass);
+
+    /**
+     * For types it would find the corresponding resolved types.
+     */
+    <T> T toResolvedType(Type javaparserType, Class<T> resultClass);
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java:39]: I feel that `resolveType` means the same thing as `toResolvedType` and is more in line with `resolveDeclaration`.

File: javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java [hunk 2]
```diff
@@ -22,7 +22,21 @@
...
+
+    ResolvedType calculateType(Expression expression);
```
Comment (matozoid) [inline javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java:41]: ... or maybe I just don't understand this very well ;-)

File: javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java:39]: I am not very good with naming, especially in English. The difference I see is that I can have a reference (es. "foo") and from that I want to find the correspinding declaration (es. "private int foo;"). This is what `resolveDeclaration` does. `toResolvedType` instead takes a JP type and give me a JSS type. Sure, to build the JSS type it does all the necessary resolving. However to me it seems more like a conversion. Making we can calling `correspondingResolvedType` or something like that.

File: javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java (hunk 2 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java:41]: ...or maybe I am just terrible at giving names :D

File: javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java (hunk 1 above)
Comment (matozoid) [reply javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java:39]: I'm mostly thinking about "discoverability" - if I type a variable and a dot in my IDE and scroll through the autosuggested methods, the methods from this interface would not show up next to each other.

But that's a small thing. Keep it as it is then :-)

**Review by matozoid — APPROVED** [review]

File: javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java (hunk 1 above)
Comment (ftomassetti) [reply javaparser-core/src/main/java/com/github/javaparser/resolution/SymbolResolver.java:39]: I could make the two methods start with get (getDeclaration, getResolvedType) or with to (toDeclaration, toResolvedType)

**Comment by ftomassetti** [comment]
Well, given it is approved let's merge it

**Pull request merged**